library/src/scala/internal/quoted/Matcher.scala

```diff
@@ -138,11 +139,28 @@ private[quoted] object Matcher {
             matched(scrutinee.seal)
 
           // Match a scala.internal.Quoted.patternHole and return the scrutinee tree
-          case (scrutinee: Term, TypeApply(patternHole, tpt :: Nil))
+          case (ClosedTerm(scrutinee), TypeApply(patternHole, tpt :: Nil))
               if patternHole.symbol == internal.Definitions_InternalQuoted_patternHole &&
                  scrutinee.tpe <:< tpt.tpe =>
             matched(scrutinee.seal)
 
+          // Matches an open term and wraps it into a lambda that provides the free variables
+          case (scrutinee, pattern @ Apply(Select(TypeApply(Ident("patternHole"), List(Inferred())), "apply"), args0 @ IdentArgs(args))) =>
```

Comment:
It seems the only sanity check is `Ident("patternHole")`, which is too weak --- in case programmers use the same name.